File: erlang/proctask.erl
-module(proctask).
-compile(export_all).


loop() -> 
    receive
        Task      -> proc(task:name(Task),Task)
        after 500 -> ok
    end,
    loop().
    

% OrdenName == userLsd, userDelete, userCreate, userOpenRead, userOpenWrite, userWrite, userRead, userClose, userBye
% OrdenWorkerName == workerDelete, workerOpenRead, workerWrite, workerRead, workerOpenWrite, workerSay, workerOpenSucc, workerClose, workerCloseBye, workerCloseSucc, workerToken

%getOwner( s: String ) = noowner | WorkerId
%myFiles (file: String) = s : NoFile | Unused | Reading | Writing

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%LSD
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
proc( userLsd , Task ) ->
    mensaje = lists:foldl( fun(C,L) ->mensaje:addArg(C,L) end, mensaje:mOk(),  globalfiles:archivosActuales() ),
    comunic:responderCliente( task:cliente(Task), mensaje ),
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Create
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc( userCreate , Task ) ->
    Name = task:fileName(Task),
    C  = task:cliente(Task),
    case globalfiles:getOwner(Name) of
         noOwner -> tokenqueues:newCreate(Name,C);
         _       -> responderClienteRemoto(C, mensaje:archivoExistente())
    end,
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Open
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Write
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc( userOpenWrite, Task)->
    Name = task:fileName(Task),
    Idg  = task:idGlobal(Task),
    C    = task:cliente(Task),
    case globalfiles:getOwner(Name) of
         noOwner -> responderClienteRemoto(Idg, mensaje:archivoNoExiste());
         W       -> Orden = task:crear_workerOpenWrite(Name, ids:makeIdGlobal(ids:myId(),C)),
                    comunic:enviarWorker(W,Orden)
    end,
    ok;

proc (workerOpenWrite, Task)->
    Name = task:fileName(Task),
    IdG  = task:idGlobal(Task),
    case localfiles:status(Name) of
         noFile  -> responderClienteRemoto(IdG, mensaje:archivoNoExiste());
         unused  -> Handle  = realfs:openw(Name),% openr dado un Name te devuelve un handle
                    LocalFd = fdmanage:registerFd(IdG,Handle),
                    localfiles:openW(Name),
                    Gfd = ids:makeGlobalFd(LocalFd, ids:myId()),
                    C   = ids:globalIdToClient(IdG),
                    W   = ids:globalIdToWorker(IdG),
                    Orden = task:crear_workerOpenSucc(Gfd, C),
                    comunic:enviarWorker(W,Orden);
         _       -> responderClienteRemoto(IdG, mensaje:archivoOcupado())
    end,
    ok;



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Read
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc( userOpenRead, Task)->
    Name = task:fileName(Task),
    Idg  = task:idGlobal(Task),
    C    = task:cliente(Task),
    case globalfiles:getOwner(Name) of
         noOwner -> responderClienteRemoto(Idg, mensaje:archivoNoExiste());
         W       -> Orden = task:crear_workerOpenRead(Name, ids:makeIdGlobal(ids:myId(),C)),
                    comunic:enviarWorker(W,Orden)
    end,
    ok;


proc (workerOpenRead, Task)->
    Name = task:fileName(Task),
    IdG  = task:idGlobal(Task),
    case localfiles:status(Name) of
         noFile  -> responderClienteRemoto(IdG, mensaje:archivoNoExiste());
         writing -> responderClienteRemoto(IdG, mensaje:archivoOcupado());
         _       -> Handle  = realfs:openr(Name),
                    LocalFd = fdmanage:registerFd(IdG,Handle),
                    localfiles:openR(Name),
                    Gfd = ids:makeGlobalFd(LocalFd, ids:myId()),
                    C   = ids:globalIdToClient(IdG),
                    W   = ids:globalIdToWorker(IdG),
                    Orden = task:crear_workerOpenSucc(Gfd, C),
                    comunic:enviarWorker(W,Orden)
    end,
    ok;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%OpenSucc
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


proc(workerOpenSucc, Task)->
    Gfd = ids:fdGlobal(Task),
    C   = ids:cliente(Task),
    openedfiles:registerOpen(Gfd,C),
    W   = ids:globalFdToWorker(Gfd),
    IdG = ids:makeIdGlobal(W,C),
    responderClienteRemoto(IdG, mensaje:archivoOcupado()),
    ok;


%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Write & Read
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Write
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc(userWrite, Task)->
    Gfd = task:fdGlobal(Task),
    C   = task:cliente(Task),
    IdG = ids:makeIdGlobal(ids:myId(),C),
    W   = ids:globalFdToWorker(Gfd),
    Orden = task:crear_workerWrite(Gfd, IdG),
    responderClienteRemoto(W,Orden),
    ok;

proc(workerWrite, Task)->
    Gfd = task:fdGlobal(Task),
    IdG = task:idGlobal(Task),
    Fd  = ids:globalFdToLocalFd(Gfd),
    case fdmanage:getOwner(Fd)==IdG of
         false -> responderClienteRemoto( IdG, mensaje:permisoDenegado());
         true  -> Handle = fdmanage:getHandle(Fd),
                  Data   = task:strTxt(Task),
                  realfs:write ( Handle, Data),
                  responderClienteRemoto( IdG, mensaje:archivoWriteSucc())
    end,
    ok;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% Read
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc(userRead, Task)->
    Gfd = task:fdGlobal(Task),
    C   = task:cliente(Task),
    Sz  = task:sizeTxt(Task),
    IdG = ids:makeIdGlobal(ids:myId(),C),
    W   = ids:globalFdToWorker(Gfd),
    Orden = task:crear_workerRead(Sz, Gfd, IdG),
    responderClienteRemoto(W,Orden),
    ok;

proc(workerRead, Task)->
    Gfd = task:fdGlobal(Task),
    IdG = task:idGlobal(Task),
    Fd  = ids:globalFdToLocalFd(Gfd),
    case fdmanage:getOwner(Fd)==IdG of
         false -> responderClienteRemoto( IdG, mensaje:permisoDenegado());
         true  -> Handle = fdmanage:getHandle(Fd),
                  Sz     = task:sizeTxt(Task),
                  Txt    = realfs:read(Handle,Sz),
                  responderClienteRemoto( IdG, mensaje:archivoReadSucc(Txt))
    end,
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Close
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc( userClose, Task)->
    Gfd   = task:fdGlobal(Task),
    C     = task:cliente(Task),
    Orden = task:crear_workerClose(Gfd, ids:makeIdGlobal(ids:myId(),C)),
    W     = ids:globalFdToWorker(Gfd),
    comunic:enviarWorker(W,Orden),
    ok;

proc( workerClose, Task ) ->
    Gfd = task:fdGlobal(Task),
    Idg = task:idGlobal(Task),
    Fd  = ids:globalFdToLocalFd(Gfd),
    case fdmanage:getOwner(Fd)==Idg of
         false -> responderClienteRemoto(Idg, mensaje:permisoDenegado());
         true  -> fdManage:unregisterFd(Fd),
                  F     = task:fileName(Task),
                  localfiles:close(F),
                  Orden = task:crear_workerCloseSucc(Gfd, Idg),
                  W     = ids:globalIdToWorker(Idg),
                  comunic:enviarWorker(W,Orden)
    end,
    ok;

proc( workerCloseSucc, Task )->
    Idg   = task:idGlobal(Task),
    C     = ids:globalIdToClient(Idg),
    Gfd   = task:fdGlobal(Task),
    openedfiles:registerClose(Gfd),
    comunic:responderCliente(C,mensaje:archivoCerrado()),
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Delete
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



proc( userDelete , Task ) ->
    Name = task:fileName(Task),
    C = task:cliente(Task),
    case globalfiles:getOwner(name) of
         noOwner -> comunic:responderCliente( c , mensaje:archivoNoExiste() ) ;
         W       -> Orden = task:crear_workerDelete(Name, ids:makeIdGlobal(ids:myId(),C)),
                    comunic:enviarWorker(W,Orden)
    end,
    ok;


proc( workerDelete , Task ) ->
    Name = task:fileName(Task),
    Idg  = task:idGlobal(Task),
    case localfiles:status(Name) of
         noFile -> responderClienteRemoto(Idg, mensaje:archivoNoExiste());
         unused -> tokenqueues:newDelete(Name),
                   responderClienteRemoto(Idg, mensaje:archivoBorrado()),
                   localfiles:delete(Name);
         _      -> responderClienteRemoto(Idg, mensaje:archivoOcupado())
    end,
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Bye
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc ( userBye, Task )->
    C        = task:cliente(Task),
    GlobalFD = openedfile:globalFdList(C),
    Fun = fun(Gfd) -> 
             Orden = task:crear_workerCloseBye(Gfd),
             W     = ids:globalFdToWorker(Gfd),
             comunic:enviarWorker(W,Orden),
             ok
          end,
    lists:foreach(Fun,GlobalFD),
    ok;

proc( workerCloseBye, Task ) ->
    Gfd = task:fdGlobal(Task),
    Fd  = ids:globalFdToLocalFd(Gfd),
    fdManage:unregisterFd(Fd),
    F     = task:fileName(Task),
    localfiles:close(F),
    openedfiles:registerClose(Gfd),
    ok;

%---------------------------------------------------------------------------------------------------------------


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Token
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% newCreate : String, ClientId -> ok (agrega a la lista de candidatos a crearse, el cliente es para saber a quien responder)
% newDelete : String -> ok (agrega a la lista de bajas a informarse)
% getCreates : () -> [{String,WiD}] (extrae la lista de creaciones candidatas)
% getDeletes : () -> [String] (extrae la lista de bajas a informarse)

% tickTime : () -> Int (devuelve el tiempo que tarda el token en ser procesado)
% recvT : Token -> ok (notifica al modulo que se recibio el Token y lo almacena)
% getT : () -> Token (devuelve el Token almacenado)
% mustProc() : () -> true | false (indica si se debe procesar o no el Token)
      %  ListaBajas = getListaBajasToken();
     %   ListaAltas = getListaAltasToken();
    %    Filtrar los que tienen WorkerId == myid() (ambas listas)
   %     Los de la cola de delete ponerlos en ListaBajas
  %      Los de mi cola de create que estene en ListaAltras contestar que no se puede
 %       Los de mi cola de create que NO estene en ListaAltras contestar que se pudo crear (antes modificar estado mi BD local)  y Agregar a lista de altas
%
 %       Modificar estado global con las listas Alta/Baja
%        pasar el token

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                    
proc(workerToken, Task) ->
    Token = task:token(Task),
    task:recvT(Token),
    ok;

%---------------------------------------------------------------------------------------------------------------

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%Say
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

proc(workerSay, Task)->
    C   = task:cliente(Task),
    Msj = task:mensaje(Task),
    comunic:responderCliente(C,Msj),    
    ok.

responderClienteRemoto( Idg, Msj) -> 
    W = ids:globalIdToWorker(Idg),
    C = ids:globalIdToClient(Idg),
    T = task:crear_workerSay(C, Msj ),
    comunic:enviarWorker(W, T).

%---------------------------------------------------------------------------------------------------------------

    

File: cpp/procTask.cpp
#include <bits/stdc++.h>
#include <string>
#include "procTask.h"

using namespace std;
#define forall(it,v) for(auto it=v.begin();it!=v.end();++it)

#define dprint(v) cerr << #v"=" << v << endl //;)
#define forr(i,a,b) for(int i=(a); i<(b); i++)
#define forn(i,n) forr(i,0,n)
#define dforn(i,n) for(int i=n-1; i>=0; i--)
#define sz(c) ((int)c.size())
#define zero(v) memset(v, 0, sizeof(v))
#define pb push_back
#define fst first
#define snd second
typedef long long ll;
typedef pair<int,int> ii;

const int MAXN=100100;
int n;

void responderClienteRemoto( GlobalId Idg,mensaje Msj){
	WorkerId W = idsManage::globalIdToWorker(Idg);
	ClientId C = idsManage::globalIdToClient(Idg);
	task T = task::crear_workerSay(C, Msj);
	enviarWorker(W,T);
}

void caseUserCon(WorkerScope *who,task& t){
				ClientId c = t.getCliente();
				responderCliente(c,mensaje::coneccionEstablecida(c),who);
}

void caseUserLsd(WorkerScope *who,task& t){
			mensaje l = mensaje::mOk();
			vector<string> arch = who->MyglobalFiles.archivosActuales();
			forall (it, arch ){ l.addArg(*it);  };
			ClientId cID = t.getCliente();
			responderCliente(cID,l, who);
}

void caseUserCreate(WorkerScope *who,task& t){
	string name = t.getFileName();
	ClientId cID = t.getCliente();
	WorkerId wID = who->MyglobalFiles.getOwner(name);
	if (noWorkerId!=wID){
		responderCliente(cID,mensaje::archivoExistente(),who);
	}else{
		if (who->MytokenQueues.isInCreate(name)) {
			responderCliente(cID,mensaje::archivoExistente(),who);
		}else{
			who->MytokenQueues.newCreate(name, cID);
		};
	}; 
}
void caseUserOpenRead(WorkerScope *who,task& t){
	string name = t.getFileName();
	ClientId cID = t.getCliente();
	GlobalId gID = idsManage::makeIdGlobal(who->MyIdsManage.myId(), cID);
	WorkerId wID = who->MyglobalFiles.getOwner(name);
	if (noWorkerId==wID){
		responderCliente(cID,mensaje::archivoNoExiste(),who);
	}else{
		task t2 = task::crear_workerOpenRead(name,gID);
		enviarWorker(wID,t2);
	}; 
}

void caseUserOpenWrite(WorkerScope *who,task& t){
	string name = t.getFileName();
	ClientId cID = t.getCliente();
	GlobalId gID = idsManage::makeIdGlobal(who->MyIdsManage.myId(), cID);
	WorkerId wID = who->MyglobalFiles.getOwner(name);
	if (noWorkerId==wID){
		responderCliente(cID,mensaje::archivoNoExiste(),who);
	}else{
		task t2 = task::crear_workerOpenWrite(name,gID);
		enviarWorker(wID,t2);
	}; 
}

void caseUserDelete(WorkerScope *who,task& t){
	string name  = t.getFileName();
	ClientId cID = t.getCliente();
	GlobalId gID = idsManage::makeIdGlobal(who->MyIdsManage.myId(), cID);
	WorkerId wID = who->MyglobalFiles.getOwner(name);
	if (noWorkerId==wID){
		responderCliente(cID,mensaje::archivoNoExiste(),who);
	}else{
		task t2 = task::crear_workerDelete(name,gID);
		enviarWorker(wID,t2);
	}; 
}

void caseUserBye(WorkerScope *who,task& t){
	ClientId cID = t.getCliente();
	vector<GlobalFd> gFds = who->MyopenedFiles.globalFdList(cID);
	responderCliente(cID,mensaje::mOk(),who);
	WorkerId w;
	task order;
	forall(it,gFds){
			order=task::crear_workerCloseBye(*it);
			w=idsManage::globalFdToWorker(*it);
			enviarWorker(w,order);
			who->MyopenedFiles.registerClose(*it);
	}
}

void caseUserWrite(WorkerScope *who,task& t){
	ClientId cID = t.getCliente();
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = idsManage::makeIdGlobal(who->MyIdsManage.myId(),cID);
	vector<GlobalFd> gFdS = (who->MyopenedFiles.globalFdList(cID));

	if ( find(gFdS.begin(), gFdS.end(), gFd) != gFdS.end() ){
		int sizeF = t.getSizeTxt();
		int sizeA = (t.getStrTxt()).size();
		string txt;
		if(sizeF<sizeA){
			txt = t.getStrTxt(); 
		}else{
			txt = t.getStrTxt().substr(0,sizeF);
		}
		WorkerId w = idsManage::globalFdToWorker(gFd);
		task order = task::crear_workerWrite(txt,gFd, idG);
		comunic:enviarWorker(w,order);
	}else{
		responderCliente(cID, mensaje::permisoDenegado(), who);
	}
}
	

void caseUserRead(WorkerScope *who,task& t){
	ClientId cID = t.getCliente();
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = idsManage::makeIdGlobal(who->MyIdsManage.myId(),cID);
	int sizeF = t.getSizeTxt();
	bool b=false;
	forall(it, (who->MyopenedFiles.globalFdList(cID))){ 
		b = b || (*it==gFd);
	}
	string txt;
	if (b){
		WorkerId w = idsManage::globalFdToWorker(gFd);
		task order = task::crear_workerRead(sizeF,gFd, idG);
		comunic:enviarWorker(w,order);
	}else{
		responderCliente(cID, mensaje::permisoDenegado(), who);
	}
}

void caseUserClose(WorkerScope *who,task& t){
	ClientId cID = t.getCliente();
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = idsManage::makeIdGlobal(who->MyIdsManage.myId(),cID);
	vector<GlobalFd> gFdS = (who->MyopenedFiles.globalFdList(cID));
	
	if ( find(gFdS.begin(), gFdS.end(), gFd) != gFdS.end() ){
		WorkerId w = idsManage::globalFdToWorker(gFd);
		task order = task::crear_workerClose(gFd, idG);
		comunic:enviarWorker(w,order);
	}else{
		responderCliente(cID, mensaje::permisoDenegado(), who);
	}
}

void caseWorkerDelete(WorkerScope *who,task& t){
	string name  = t.getFileName();
	GlobalId idG = t.getIdGlobal();
	fileStatus state = who->MylocalFiles.status(name);
	if (state == noFile){
		responderClienteRemoto(idG,mensaje::archivoNoExiste());
	}else if (state == unused){
		who->MytokenQueues.newDelete(name);
		responderClienteRemoto(idG, mensaje::archivoBorrado());
		who->MylocalFiles.deletef(name);
		deletef(name);
	}else{
		responderClienteRemoto(idG, mensaje::archivoOcupado());
	}
}
void caseWorkerOpenRead(WorkerScope *who,task& t){
	string name  = t.getFileName();
	GlobalId idG = t.getIdGlobal();
	fileStatus state = who->MylocalFiles.status(name);
	
	if (state == noFile){
		responderClienteRemoto(idG,mensaje::archivoNoExiste());
	}else if (state == writing){
		responderClienteRemoto(idG, mensaje::archivoOcupado());
	}else{
		RealFSHandle handle = openr(name);
		LocalFd localFd = who->MyfdManage.registerFd(idG, name, handle);
		who->MylocalFiles.openR (name);
		WorkerId wID = who->MyIdsManage.myId();
		GlobalFd gFd = idsManage::makeGlobalFd(localFd, wID);
		ClientId  c  = idsManage::globalIdToClient(idG);
		WorkerId  w  = idsManage::globalIdToWorker(idG);
		task order   = task::crear_workerOpenSucc(gFd, c);
		enviarWorker(w, order);		
	}
}
void caseWorkerOpenWrite(WorkerScope *who,task& t){
	string name  = t.getFileName();
	GlobalId idG = t.getIdGlobal();
	fileStatus state = who->MylocalFiles.status(name);
	if (state == noFile){
		responderClienteRemoto(idG,mensaje::archivoNoExiste());
	}else if (state != unused){
		responderClienteRemoto(idG, mensaje::archivoOcupado());
	}else{
		RealFSHandle handle = openw(name);
		LocalFd localFd = who->MyfdManage.registerFd(idG, name, handle);
		who->MylocalFiles.openW(name);
		WorkerId wID = who->MyIdsManage.myId();
		GlobalFd gFd = idsManage::makeGlobalFd(localFd, wID);
		ClientId  c  = idsManage::globalIdToClient(idG);
		WorkerId  w  = idsManage::globalIdToWorker(idG);
		task order   = task::crear_workerOpenSucc(gFd, c);
		enviarWorker(w, order);		
	}
}
void caseWorkerOpenSucc(WorkerScope *who,task& t){
	ClientId c = t.getCliente();
	GlobalFd gFd = t.getGlobalFd();
	who->MyopenedFiles.registerOpen(gFd,c);
	responderCliente(c, mensaje::archivoAbierto(gFd), who);
}
void caseWorkerRead(WorkerScope *who,task& t){
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = t.getIdGlobal();
	LocalFd fd = idsManage::globalFdToLocalFd(gFd);
	if (who->MyfdManage.getOwner(fd) != idG){
		responderClienteRemoto(idG,mensaje::permisoDenegado());
	}else{
		RealFSHandle handle = who->MyfdManage.getHandle(fd);
		int sz = t.getSizeTxt();
		string txt;
		if(read(handle,sz,txt)){
			responderClienteRemoto( idG, mensaje::archivoReadSucc(txt));
		}else{
			responderClienteRemoto( idG, mensaje::finDeArchivo());
		}
	}
}
void caseWorkerWrite(WorkerScope *who,task& t){
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = t.getIdGlobal();
	LocalFd fd = idsManage::globalFdToLocalFd(gFd);
	if (who->MyfdManage.getOwner(fd) != idG){
		responderClienteRemoto(idG,mensaje::permisoDenegado());
	}else{
		RealFSHandle handle = who->MyfdManage.getHandle(fd);
		string txt = t.getStrTxt();
		write(handle, txt);
		responderClienteRemoto( idG, mensaje::archivoWriteSucc());
		
	}
}
void caseWorkerSay(WorkerScope *who,task& t){
	ClientId c = t.getCliente();
	mensaje msj = t.getMensaje();
	responderCliente(c, msj, who);
	
}

void caseWorkerClose(WorkerScope *who,task& t){
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = t.getIdGlobal();
	LocalFd fd = idsManage::globalFdToLocalFd(gFd);
	if (who->MyfdManage.getOwner(fd) != idG){
		responderClienteRemoto(idG,mensaje::permisoDenegado());
	}else{
		string file = who->MyfdManage.getNameFile(fd);
		who->MyfdManage.unregisterFd(fd);
		who->MylocalFiles.close(file);
		task order = task::crear_workerCloseSucc(gFd, idG);
		WorkerId  w  = idsManage::globalIdToWorker(idG);
		enviarWorker(w,order);
	}
}
void caseWorkerCloseBye(WorkerScope *who,task& t){
	GlobalFd gFd = t.getGlobalFd();
	LocalFd fd = idsManage::globalFdToLocalFd(gFd);
	string file = who->MyfdManage.getNameFile(fd);
	who->MyfdManage.unregisterFd(fd);
	who->MylocalFiles.close(file);
	who->MyopenedFiles.registerClose(gFd);
}
void caseWorkerCloseSucc(WorkerScope *who,task& t){
	GlobalFd gFd = t.getGlobalFd();
	GlobalId idG = t.getIdGlobal();
	ClientId cID = idsManage::globalIdToClient(idG);
	who->MyopenedFiles.registerClose(gFd);
	responderCliente(cID,mensaje::archivoCerrado(), who);
}
void caseWorkerToken(WorkerScope *who,task& t){
	token tk = t.getToken();
	who->MytokenControl.recvT(tk);
}

	
void procTask(WorkerScope *who,task& t) {
	switch (t.getTaskName()) {
		case userCon:
			caseUserCon(who,t); 
		break;
		case userLsd:
			caseUserLsd(who,t);
		break;
		case userCreate:
			caseUserCreate(who,t);
		break;
		case userOpenRead:
			caseUserOpenRead(who,t);
		break;
		case userOpenWrite:
			caseUserOpenWrite(who,t);
		break;
		case userDelete:
			caseUserDelete(who,t);
		break;
		case userWrite:
			caseUserWrite(who,t);
		break;
		case userRead:
			caseUserRead(who,t);
		break;
		case userClose:
			caseUserClose(who,t);
		break;
		case userBye:
			caseUserBye(who,t);
		break;
		case workerDelete:
			caseWorkerDelete(who,t);
		break;
		case workerOpenRead:
			caseWorkerOpenRead(who,t);
		break;
		case workerOpenWrite:
			caseWorkerOpenWrite(who,t);
		break;
		case workerRead:
			caseWorkerRead(who,t);
		break;
		case workerWrite:
			caseWorkerWrite(who,t);
		break;
		case workerSay:
			caseWorkerSay(who,t);
		break;
		case workerOpenSucc:
			caseWorkerOpenSucc(who,t);
		break;
		case workerClose:
			caseWorkerClose(who,t);
		break;
		case workerCloseBye:
			caseWorkerCloseBye(who,t);
		break;
		case workerCloseSucc:
			caseWorkerCloseSucc(who,t);
		break;
		case workerToken:
			caseWorkerToken(who,t);
		break;
	}
}
